lib/urlwatch/reporters.py

```diff
@@ -668,18 +668,75 @@ def submit(self):
+        # message is too long.
+        remaining_len = max_length - summary_len - 40
+
+        if summary and show_footer:
+            footer = ('--- ',
+                      '%s %s, %s  ' % (urlwatch.pkgname, urlwatch.__version__, urlwatch.__copyright__),
+                      'Website: %s  ' % (urlwatch.__url__,),
+                      'watched %d URLs in %d seconds' % (len(self.job_states), self.duration.seconds))
+            footer_len = sum(len(part) for part in footer) + len(footer) - 1
+            remaining_len -= footer_len
+
+        # Calculate approximate available length per item, shared equally.
+        len_per_details = remaining_len // len(details)
+
         if summary:
             yield from ('%d. %s' % (idx + 1, line) for idx, line in enumerate(summary))
             yield ''
 
         if show_details:
-            yield from details
+            if len_per_details < 100:
+                yield "Details omitted due to message length."
+            else:
+                unprocessed = len(details)
+
+                for header, body in details:
+                    # Calculate the available length for the body and render it
+                    avail_length = len_per_details - len(header) - 1
+                    body = MarkdownReporter._format_details_body(body, avail_length)
+                    actual_length = len(header) + len(body)
+                    unprocessed -= 1
+
+                    yield header
+                    yield body
+                    yield ''
+
+                    # Distribute the unused length into subsequent items,
+                    # unless we're at the last item already.
+                    unused = len_per_details - actual_length
+                    remaining_len -= len_per_details
+                    remaining_len += unused
+
+                    if unprocessed > 0:
+                        len_per_details = remaining_len // unprocessed
 
         if summary and show_footer:
-            yield from ('--- ',
-                        '%s %s, %s  ' % (urlwatch.pkgname, urlwatch.__version__, urlwatch.__copyright__),
-                        'Website: %s  ' % (urlwatch.__url__,),
-                        'watched %d URLs in %d seconds' % (len(self.job_states), self.duration.seconds))
+            yield from footer
+
+    @staticmethod
+    def _format_details_body(s, max_length):
+        # Number of characters for the Markdown code block wrapper (three
+        # backticks and a newline, twice).
+        wrapper_length = 8
+
+        # Message to print when the diff is too long.
+        trim_message = "[... diff trimmed ...]"
+        trim_message_length = len(trim_message)
+
+        if max_length is None or len(s) + wrapper_length <= max_length:
+            return "```diff\n{}\n```".format(s)
+        else:
+            while len(s) + wrapper_length + trim_message_length > max_length:
+                s = re.sub("\n.*$", "", s)
```

Comment:
What if there's no newline in the string anymore (e.g. because there's a really really long line somewhere), would this be an infinite loop?

Reply:
Oops, you're right. I modified this so that this problem cannot happen anymore.